### PR TITLE
MGMT-1479 Add stage_started_at and stage_updated_at & MGMT-1493 fix s…

### PR DESCRIPTION
**Pull request by YuviGold**
…tatus_updated_at

[MGMT-1493](https://issues.redhat.com/browse/MGMT-1493) Fixing `status_updated_at`
[MGMT-1479](https://issues.redhat.com/browse/MGMT-1479) Adding `stage_started_at` and `stage_updated_at` fields
[MGMT-1495](https://issues.redhat.com/browse/MGMT-1495) Fixing `UpdateClusterStatus` and `UpdateHostStatus` functions confusion and duplication

### CODE REVIEW

File: internal/cluster/common.go [hunk 1]
```diff
@@ -41,52 +41,59 @@ type baseState struct {
...
+	// Status is queried as well to avoid races between different components.
+	dbReply := db.Model(&common.Cluster{}).Where("id = ? and status = ?",
+		c.ID.String(), swag.StringValue(c.Status)).Updates(updates)
 	if dbReply.Error != nil {
 		return errors.Wrapf(dbReply.Error, "failed to update cluster %s state from %s to %s",
-			c.ID.String(), srcStatus, swag.StringValue(c.Status))
+			c.ID.String(), swag.StringValue(c.Status), status)
 	}
-	if dbReply.RowsAffected == 0 && swag.StringValue(c.Status) != srcStatus {
+	if dbReply.RowsAffected == 0 {
```
Comment (ronniel1) [inline internal/cluster/common.go:86]: There is a behaivor change here.
In the new version, if the old-status == new-status there will be an error. in the old version, there will not be an error

File: internal/cluster/common.go [hunk 2]
```diff
@@ -41,52 +41,59 @@ type baseState struct {
...
 		return errors.Errorf("failed to update cluster %s state from %s to %s, nothing have changed",
-			c.ID.String(), srcStatus, swag.StringValue(c.Status))
+			c.ID.String(), swag.StringValue(c.Status), status)
 	}
-	c.StatusInfo = &statusInfo
-	log.Infof("Updated cluster <%s> status from <%s> to <%s> with fields: %s",
-		c.ID.String(), srcStatus, swag.StringValue(c.Status), updates)
+	log.Infof("updated cluster %s from state <%s> to state <%s>", c.ID.String(), swag.StringValue(c.Status), status)
+
+	if err := db.First(c, "id = ?", c.ID.String()).Error; err != nil {
```
Comment (ronniel1) [inline internal/cluster/common.go:92]: This is another change from before, c passed as an argument is being changed
This has a far-reaching affect, since in most cases this c argument is passed along to many places
Seems kind of dangerous

File: internal/cluster/common.go [hunk 3]
```diff
@@ -41,52 +41,59 @@ type baseState struct {
...
 	return &UpdateReply{
-		State:     state,
-		IsChanged: state != swag.StringValue(c.Status),
+		State:     status,
+		IsChanged: wouldChange,
 	}, nil
 }
 
-func updateClusterStateWithParams(log logrus.FieldLogger, srcStatus, statusInfo string, c *common.Cluster, db *gorm.DB,
+func updateClusterStateWithParams(log logrus.FieldLogger, status string, statusInfo string, c *common.Cluster, db *gorm.DB,
```
Comment (filanov) [inline internal/cluster/common.go:57]: I'm confused. state machine change the status of the object `common.Cluster`
if `status` is the new status and `common.Cluster` already have the new status then `tatus != swag.StringValue(c.Status)` will be always the same no?

File: internal/cluster/common.go [hunk 4]
```diff
@@ -41,52 +41,59 @@ type baseState struct {
...
 	extra ...interface{}) error {
 
-	updates := map[string]interface{}{"status": swag.StringValue(c.Status), "status_info": statusInfo}
+	updates := map[string]interface{}{"status": status, "status_info": statusInfo}
+
+	if status != swag.StringValue(c.Status) {
+		updates["status_updated_at"] = strfmt.DateTime(time.Now())
+	}
+	if *c.Status == clusterStatusReady && status == clusterStatusPrepareForInstallation {
```
Comment (filanov) [inline internal/cluster/common.go:65]: please use `models.ClusterStatusPrepareForInstallation`

Comment (filanov) [reply internal/cluster/common.go:65]: The same for all the other states in this function.

File: internal/cluster/common.go (hunk 1 above)
Comment (filanov) [inline internal/cluster/common.go:86]: what will happen if the everything stay the same? this function will fail no?

File: internal/cluster/transition.go [hunk 5]
```diff
@@ -63,6 +64,14 @@ func (th *transitionHandler) PostResetCluster(sw stateswitch.StateSwitch, args s
...
-	return updateClusterStateWithParams(logutil.FromContext(params.ctx, th.log), sCluster.srcState,
+	return updateCluster(logutil.FromContext(params.ctx, th.log), sCluster.srcState,
 		params.reason, sCluster.cluster, params.db)
 }
+
+// Updates the status according to the cluster where the status equals the srcStatus
+func updateCluster(log logrus.FieldLogger, srcStatus, statusInfo string, c *common.Cluster, db *gorm.DB,
+	extra ...interface{}) error {
+	newStatus := c.Status
+	c.Status = &srcStatus
```
Comment (filanov) [inline internal/cluster/transition.go:75]: confusing.

File: internal/cluster/common.go (hunk 3 above)
Comment (ronniel1) [inline internal/cluster/common.go:57]: Why did you change the semantics here?
Before it was a change from the srcStatus to c.status
Now its a change from c.status to status

Comment (YuviGold) [reply internal/cluster/common.go]: The naming were changed in order to make it clearer. Please look at the final changes.

File: internal/cluster/common.go (hunk 2 above)
Comment (YuviGold) [reply internal/cluster/common.go]: Changed to return a value instead of changing an object.

File: internal/cluster/common.go (hunk 1 above)
Comment (YuviGold) [reply internal/cluster/common.go]: Fixed.

File: internal/cluster/transition.go (hunk 5 above)
Comment (YuviGold) [reply internal/cluster/transition.go]: Removed.

File: internal/cluster/common.go (hunk 1 above)
Comment (YuviGold) [reply internal/cluster/common.go]: This part was split into two functions - `UpdateCluster` verifies the DB behavior. While `UpdateClusterStatus` verifies that the  status has changed.

File: internal/cluster/common.go (hunk 4 above)
Comment (YuviGold) [reply internal/cluster/common.go]: Fixed.

File: internal/cluster/common.go (hunk 3 above)
Comment (YuviGold) [reply internal/cluster/common.go]: Removed the Cluster struct to avoid confusion.

File: internal/cluster/common.go [hunk 6]
```diff
@@ -32,66 +32,58 @@ const (
...
-	dbReply := db.Model(&models.Cluster{}).Where("id = ? and status = ?", c.ID.String(), srcStatus).Updates(updates)
-	if dbReply.Error != nil {
-		return errors.Wrapf(dbReply.Error, "failed to update cluster %s state from %s to %s",
-			c.ID.String(), srcStatus, swag.StringValue(c.Status))
+
+	// Query by <cluster-id, status>
+	// Status is required as well to avoid races between different components.
+	dbReply := db.Model(&common.Cluster{}).Where("id = ? and status = ?", clusterId, srcStatus).Updates(updates)
+
+	if dbReply.Error != nil || dbReply.RowsAffected == 0 {
```
Comment (filanov) [inline internal/cluster/common.go:74]: Not sure if we should add `swag.StringValue(c.Status) != srcStatus` validation as well. 
if all the tests passed then maybe it's ok not to add it. 
but i think that we may have a state that we try to update and leave the same state without any changes, and it will be bad if fails.

Comment (YuviGold) [reply internal/cluster/common.go:74]: I don't know about the swagger casting. 
Testing works fine.
Why would you want to allow "Update" without any changes?

Comment (filanov) [reply internal/cluster/common.go:74]: if all tests pass (including test-infa) then it's fine

File: internal/host/known.go [hunk 7]
```diff
@@ -29,6 +29,6 @@ type knownState struct {
 	connectivityValidator connectivity.Validator
 }
 
-func (k *knownState) RefreshStatus(ctx context.Context, h *models.Host, db *gorm.DB) (*UpdateReply, error) {
+func (k *knownState) RefreshStatus(ctx context.Context, h *models.Host, db *gorm.DB) (*models.Host, error) {
```
Comment (ori-amizur) [inline internal/host/known.go:32]: Why is the change in the return value required?

Comment (YuviGold) [reply internal/host/known.go:32]: `Monitoring` functions (cluster, host) checks and notifies the user if states were changed.

File: internal/cluster/transition.go [hunk 8]
```diff
@@ -129,6 +137,18 @@ type TransitionArgsHandlePreInstallationError struct {
...
+}
+
+func (th *transitionHandler) updateTransitionCluster(log logrus.FieldLogger, db *gorm.DB, state *stateCluster,
+	statusInfo string, extra ...interface{}) error {
+
+	if cluster, err := updateClusterStatus(log, db, *state.cluster.ID, state.srcState,
+		*state.cluster.Status, statusInfo, extra...); err != nil {
+		return err
+	} else {
+		state.cluster = cluster
```
[the remark this answers is not in the document]
Comment (YuviGold) [reply internal/cluster/transition.go:151]: Kept the previous behavior of updating the instance. I'm not sure if it is needed.

**Comment by gamli75** [comment]
@YuviGold @avishayt why stage_started_at and stage_updated_at fields are not part of the models.HostProgress ?
